hbase-common/src/main/java/org/apache/hadoop/hbase/io/DelegatingInputStream.java

```diff
@@ -0,0 +1,99 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hbase.io;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+import org.apache.yetus.audience.InterfaceAudience;
+import org.apache.yetus.audience.InterfaceStability;
+
+/**
+ * An input stream that delegates all operations to another input stream.
+ * The delegate can be switched out for another at any time but to minimize the
+ * possibility of violating the InputStream contract it would be best to replace
+ * the delegate only once it has been fully consumed. <p> For example, a
+ * ByteArrayInputStream, which is implicitly bounded by the size of the underlying
+ * byte array can be converted into an unbounded stream fed by multiple instances
+ * of ByteArrayInputStream, switched out one for the other in sequence.
+ */
+@InterfaceAudience.Private
+@InterfaceStability.Evolving
+public class DelegatingInputStream extends InputStream {
```

Comment:
Where is `FilterInputStream`?

Reply:
It is inside JDK.

java.io.FilterInputStream

Reply:
Found it, but the `FilterInputStream` constructor is not public and there's no API for switching the delegate either here or in implementing classes that are not abstract.

Reply:
Maybe you could just extend it and make a public constructor? The `in` field is protected so you can introduce get and set method in sub classes to access it. This will make your DelegatingInputStream smaller.

Reply:
I think @Apache9 is right, just extend FIS and add setDelegate (that makes tit small).

Reply:
Sorry, @bharathv helped me see I did not understand your feedback here.